forte/data/readers/html_reader.py

```diff
@@ -0,0 +1,234 @@
+            elif startswith("&#", i):
+                match = charref.match(rawdata, i)
+                if match:
+                    name = match.group()[2:-1]
+                    self.handle_charref(name)
+                    k = match.end()
+                    if not startswith(';', k - 1):
+                        k = k - 1
+                    i = self.updatepos(i, k)
+                    continue
+                else:
+                    if ";" in rawdata[i:]:  # bail by consuming &#
+                        self.handle_data(rawdata[i:i + 2])
+                        i = self.updatepos(i, i + 2)
+                    break
+            elif startswith('&', i):
+                match = entityref.match(rawdata, i)
+                if match:
+                    name = match.group(1)
+                    self.handle_entityref(name)
+                    k = match.end()
+                    if not startswith(';', k - 1):
+                        k = k - 1
+                    i = self.updatepos(i, k)
+                    continue
+                match = incomplete.match(rawdata, i)
+                if match:
+                    # match.group() will contain at least 2 chars
+                    if end and match.group() == rawdata[i:]:
+                        k = match.end()
+                        if k <= i:
+                            k = n
+                        i = self.updatepos(i, i + 1)
+                    # incomplete
+                    break
+                elif (i + 1) < n:
+                    # not the end of the buffer, and can't be confused
+                    # with some other construct
+                    self.handle_data("&")
+                    i = self.updatepos(i, i + 1)
+                else:
+                    break
+            else:
+                assert 0, "interesting.search() lied"
+        # end while
+        if end and i < n and not self.cdata_elem:
+            if self.convert_charrefs and not self.cdata_elem:
+                self.handle_data(unescape(rawdata[i:n]))
+            else:
+                self.handle_data(rawdata[i:n])
+            i = self.updatepos(i, n)
+        # pylint: disable=attribute-defined-outside-init
+        self.rawdata = rawdata[i:]
+
+
+class HTMLReader(StringReader):
+    """
+    :class:`HTMLReader` is designed to read in list of html strings.
+
+    It takes in list of html strings, cleans the HTML tags and stores the
+    cleaned text in pack.
+    """
+    # pylint: disable=no-self-use,unused-argument
+    def text_replace_operation(self, text: str):
+        """Replace html tag locations with blank string.
+
+        Args:
+            text: The original html text to be cleaned.
+
+        Returns: List[Tuple[Span, str]]: the replacement operations
+
+        """
+        parser = ForteHTMLParser()
+        parser.feed(text)
+
+        return parser.spans
+
+    def _cache_key_function(self, collection):
+        return str(hash(collection)) + '.txt'
```

Comment:
Hashing on the data content could be inefficient. We hope the _cache_key_function to be as simple as possible. This is also related to the fact you are extending a StringReader. If it takes a directory of html documents, then the cache_key can simply be those paths or file names, like here: `https://github.com/asyml/forte/blob/master/forte/data/readers/plaintext_reader.py`

Further, I think the `_cache_key_function` in StringReader is wrong. Either we can change that to "hash", or we shouldn't use caching there.

Reply:
Right.. I'll update the HTMLReader to read from files and yeah.. i agree..i'll make the key_func as hash..some caching is better than no caching :)